arrow-json/src/raw/mod.rs

```diff
@@ -0,0 +1,502 @@
+                    TapeElement::StartObject(end) => end,
+                    _ => unreachable!("corrupt tape"),
+                };
+                std::mem::replace(&mut next_object, end + 1)
+            })
+            .collect();
+
+        let decoded = self.decoder.decode(&tape, &pos)?;
+        self.tape.clear();
+
+        // Sanity check
+        assert!(matches!(decoded.data_type(), DataType::Struct(_)));
+        assert_eq!(decoded.null_count(), 0);
+        assert_eq!(decoded.len(), pos.len());
+
+        // Clear out buffer
+        let columns = decoded
+            .child_data()
+            .iter()
+            .map(|x| make_array(x.clone()))
+            .collect();
+
+        let batch = RecordBatch::try_new(self.schema.clone(), columns)?;
+        Ok(Some(batch))
+    }
+}
+
+trait ArrayDecoder: Send {
+    fn decode(&mut self, tape: &Tape<'_>, pos: &[u32]) -> Result<ArrayData, ArrowError>;
+}
+
+macro_rules! primitive_decoder {
+    ($t:ty, $data_type:expr) => {
+        Ok(Box::new(PrimitiveArrayDecoder::<$t>::new($data_type)))
+    };
+}
+
+fn make_decoder(
+    data_type: DataType,
+    is_nullable: bool,
+) -> Result<Box<dyn ArrayDecoder>, ArrowError> {
+    downcast_integer! {
+        data_type => (primitive_decoder, data_type),
+        DataType::Float32 => primitive_decoder!(Float32Type, data_type),
+        DataType::Float64 => primitive_decoder!(Float64Type, data_type),
+        DataType::Boolean => Ok(Box::<BooleanArrayDecoder>::default()),
+        DataType::Utf8 => Ok(Box::<StringArrayDecoder::<i32>>::default()),
+        DataType::LargeUtf8 => Ok(Box::<StringArrayDecoder::<i64>>::default()),
+        DataType::List(_) => Ok(Box::new(ListArrayDecoder::<i32>::new(data_type, is_nullable)?)),
+        DataType::LargeList(_) => Ok(Box::new(ListArrayDecoder::<i64>::new(data_type, is_nullable)?)),
+        DataType::Struct(_) => Ok(Box::new(StructArrayDecoder::new(data_type, is_nullable)?)),
+        DataType::Binary | DataType::LargeBinary | DataType::FixedSizeBinary(_) => {
+            Err(ArrowError::JsonError(format!("{} is not supported by JSON", data_type)))
+        }
+        d => Err(ArrowError::NotYetImplemented(format!("Support for {} in JSON reader", d)))
+    }
+}
+
+fn tape_error(d: TapeElement, expected: &str) -> ArrowError {
+    ArrowError::JsonError(format!("expected {expected} got {d}"))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::reader::infer_json_schema;
+    use crate::ReaderBuilder;
+    use arrow_array::cast::{
+        as_boolean_array, as_largestring_array, as_list_array, as_primitive_array,
+        as_string_array, as_struct_array,
+    };
+    use arrow_array::types::Int32Type;
+    use arrow_array::Array;
+    use arrow_schema::{DataType, Field, Schema};
+    use std::fs::File;
+    use std::io::{BufReader, Cursor, Seek};
+    use std::sync::Arc;
+
+    fn do_read(buf: &str, batch_size: usize, schema: SchemaRef) -> Vec<RecordBatch> {
```

Comment:
I almost missed this coverage:

```suggestion
     /// decode the string as JSON with various chunk sizes to check buffer refill boundary conditions
    fn do_read(buf: &str, batch_size: usize, schema: SchemaRef) -> Vec<RecordBatch> {
```